### PR TITLE
uwsgi-cgi: adds patch to fix #10134

### LINKED ISSUE #10134: uswgi: uwsgi-cgi-luci-support broken with latest LuCI
Maintainer: @Ansuel 
Environment: -

Description:
The `uwsgi-cgi-luci-support` package is unable to execute CGI helpers in `/www/cgi-bin/`, breaking firmware upload and backup download on LuCI master.

Please refer to https://github.com/openwrt/luci/issues/3140 for a reproducer.

### PR DESCRIPTION
**Pull request by Ansuel**
- adds uwsgi patch to add option to don't follow simbolic link but call it directly (waiting to be approved)
- adds cgi-safe config option for cgi-io
- enable new cgi-dontresolve option

Signed-off-by: Ansuel Smith <ansuelsmth@gmail.com>

### CODE REVIEW

File: net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+ 	int optimize;
++	int dontresolve;
+ 	int from_docroot;
+ 	int has_mountpoints;
+ 	struct uwsgi_dyn_dict *default_cgi;
+@@ -75,6 +76,8 @@ struct uwsgi_option uwsgi_cgi_options[] = {
+ 
+         {"cgi-safe", required_argument, 0, "skip security checks if the cgi file is under the specified path", uwsgi_opt_add_string_list, &uc.cgi_safe, 0},
+ 
++		{"cgi-dontresolve", no_argument, 0 , "call simbolic link directly instead of the real path", uwsgi_opt_true,&uc.dontresolve, 0},
```
Comment (dengqf6) [inline net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option]: simbolic => symbolic

File: net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
++
+         {0, 0, 0, 0, 0, 0, 0},
+ 
+ };
+@@ -475,6 +478,7 @@ static int uwsgi_cgi_request(struct wsgi_request *wsgi_req) {
+ 
+ 	char full_path[PATH_MAX];
+ 	char tmp_path[PATH_MAX];
++	char simbolic_path[PATH_MAX];
```
Comment (dengqf6) [inline net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option]: Ditto

File: net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+ 	struct stat cgi_stat;
+ 	int need_free = 0;
+ 	int is_a_file = 0;
+@@ -533,6 +537,10 @@ static int uwsgi_cgi_request(struct wsgi_request *wsgi_req) {
+ 			uwsgi_404(wsgi_req);
+ 			return UWSGI_OK;
+ 		}
++		if (uc.dontresolve) {
++			full_path_len = strlen(full_path);
++			memcpy(simbolic_path, full_path, full_path_len+1);
```
Comment (dengqf6) [inline net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option]: Ditto

File: net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option [hunk 4]
```diff
@@ -0,0 +1,65 @@
...
+ 
+ 		full_path_len = strlen(tmp_path);
+ 		// add +1 to copy the null byte
+@@ -639,6 +647,11 @@ static int uwsgi_cgi_request(struct wsgi_request *wsgi_req) {
+ 		}
+ 	}
+ 
++	if (uc.dontresolve) {
++		full_path_len = strlen(simbolic_path);
++		memcpy(full_path, simbolic_path, full_path_len+1);
```
Comment (dengqf6) [inline net/uwsgi-cgi/030-plugins-cgi_adds_dontresolve_option]: Ditto

File: net/uwsgi-cgi/files-luci-support/uwsgi.conf [hunk 5]
```diff
@@ -25,3 +25,5 @@ die-on-term	= true
 threads		= 3
 processes	= 3
 chmod-socket	= 666
+cgi-safe = /usr/libexec/cgi-io
+cgi-dontresolve = true
```
Comment (dengqf6) [inline net/uwsgi-cgi/files-luci-support/uwsgi.conf]: Missing new line at the end of file

**Comment by Ansuel** [comment]
@danak6jq should be good now...

**Comment by danak6jq** [comment]
> @danak6jq should be good now...

I appreciate that - but I'm not sure why I'm referenced here :-)

**Comment by Ansuel** [comment]
@danak6jq erm.... sorry

@dengqf6

**Comment by peter-stadler** [comment]
The luci-cgi_io.ini should include `cgi-close-stdin-on-eof = true` for solving the read() problem in cgi-upload, cf. https://github.com/openwrt/packages/pull/10191#discussion_r333438290

**Comment by peter-stadler** [comment]
If we let luci-webui.ini die-on-idle, should we spawn new workers on demand, too? 
```
cheaper-algo = spare
cheaper = 1
cheaper-initial = 1
cheaper-step = 1
```
BTW: Why are you commenting the logger?

**Comment by peter-stadler** [comment]
I would continue to use the master process for the vassals (not the emperor), so add to both vassals:
```
master = true
idle = 900
```
Should we use `lazy-apps = true` if there is only one worker as for luci-cgi_io.ini?

**Comment by peter-stadler** [comment]
NB: I am also not sure if we should use the emperor. Its main purpose is to restart applications whose configuration ini file has changed. In openwrt that is not the default behavior: normally you have to issue a `service ... reload`. A similar behaviour could be achieved by the following uwsgi.ini (that servers both cgi-web and cgi_io, so only one socket is needed):
```
[uwsgi]
strict                  = true
socket                  = /var/run/uwsgi.sock
cgi-mode                = true
cgi                     = /www/
chdir                   = /usr/lib/lua/luci/
master                  = true
buffer-size             = 10000
reload-mercy            = 8
max-requests            = 2000
limit-as                = 200
reload-on-as            = 256
reload-on-rss           = 192
no-orphans              = true
vacuum                  = true
post-buffering          = 8192
socket-timeout          = 120
thunder-lock            = true
logger                  = syslog:uwsgi
disable-logging         = true
log-format              = %(addr) %(method) %(uri) => generated %(rsize) bytes in %(msecs) msecs
pidfile                 = /var/run/uwsgi.pid
die-on-term             = true
processes               = 8
chmod-socket            = 666
cgi-safe                = /usr/libexec/cgi-io
cgi-close-stdin-on-eof = true
cheap                   = true
cheaper-algo            = spare
cheaper                 = 1
cheaper-initial         = 1
cheaper-step            = 1
cgi-dontresolve         = true
```

**Comment by Ansuel** [comment]
Use of emperor in this case is used to suspend services and not waste
resources
Don't really understand why the master option


Il Gio 10 Ott 2019, 13:31 peter-stadler <notifications@github.com> ha
scritto:

**Comment by peter-stadler** [comment]
Without master the reload on ini changes will not work (this would be strange when using the emperor). We can also suspend services and start it on demand by: 
```
cheap                   = true
cheaper-algo            = spare
cheaper                 = 1
cheaper-initial         = 1
cheaper-step            = 1
idle                    = 900
```

**Comment by Ansuel** [comment]
@peter-stadler thx for the suggestions

i disabled logging because it's really debug log... luci output is not related to it
idle is already set by emperor on any vassal
cheaper is set only on webui as cgi-io is set to have only one worker

**Comment by peter-stadler** [comment]
For logging, we could also limit it to match regexp via log-route. This needs pcre support in uwsgi, I enabled it in the plugin version of uwsgi, see https://github.com/openwrt/packages/pull/9855

I thought maybe its better to set idle in the vassals instead of the emperor, it could be problematic if someone is adding another vassal and is not overwriting the idle time?

You are right, with the emperor cheaper makes only sense for webui. I was already thinking without emperor. (For etesync I thought about using the emperor, too. But in the end I saw no real benefit. For etesync I have the problem that it is too heavy to be started on demand. I ended up with a master for restarting it on a failure and one worker.)

**Comment by Ansuel** [comment]
i miss the first comment about emperor so i still need to add change to sites and ini files...

Anyway we are using emperor for to put cgi in idle and also to have only one uwsgi istance instead of 2 
(one for cgi-io and one for webui)

You are right about idle... 
About pcre... we canenable it by default but i think is a bit overkill use it for loggin

**Comment by Ansuel** [comment]
@peter-stadler now this should be the final version... any comment?

**Comment by peter-stadler** [comment]
pcre would not be only for logging, I mainly activated it because else there is a warning in the logs and the uwsgi part is rather small (libpcre is already plugged in by nginx if I am right).

Should we install the emperor already with the uwsgi instead of the luci support package? 
(btw: luci-*.ini should go to /etc/uwsgi/vassals)

**Comment by Ansuel** [comment]
i always lost something in the push... 

we will enable pcre part when we will push the generic uwsgi commit ok ?

**Comment by peter-stadler** [comment]
Looks good to me :-) The dontresolve is a bit of a hack, but I don't know a better solution (without changing cgi-io). Nice work.

**Comment by peter-stadler** [comment]
Yes the pcre can wait, it is not that important :-)

**Comment by Ansuel** [comment]
Anyway i think that with this is very easy and clear the switch to generic uwsgi as half the work is done
(generalize luci as a separate app)

**Comment by peter-stadler** [comment]
I am not sure, but maybe there is a fix for making cgi-io work in multi-threaded uwsgi by adding the following flag to splice:
```
index ca1575842..fd4424cb5 100644
--- a/net/cgi-io/src/main.c
+++ b/net/cgi-io/src/main.c
@@ -783,7 +783,7 @@ main_backup(int argc, char **argv)
                fflush(stdout);
 
                do {
-                       len = splice(fds[0], NULL, 1, NULL, 4096, SPLICE_F_MORE);
+                       len = splice(fds[0], NULL, 1, NULL, 4096, SPLICE_F_MORE | SPLICE_F_NONBLOCK);
                } while (len > 0);
 
                waitpid(pid, &status, 0);
```

**Comment by Ansuel** [comment]
did some test but the result was that sometimes it didn't download the entire file... will do another test with the new implementation...

**Comment by peter-stadler** [comment]
You are right, SPLICE_F_NONBLOCK is not working. On the other hand your implementation works fine. One little thing: the 030-… misses the `.patch` extension.

**Comment by Ansuel** [comment]
@hnyman can this be merged?

**Pull request merged**